filebeat/beater/filebeat.go

```diff
@@ -130,14 +144,18 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	crawler.Stop()
 
 	timeout := fb.config.ShutdownTimeout
-	if timeout > 0 {
-		logp.Info("Shutdown output timer started. Waiting for max %v.", timeout)
-
-		// Wait for either timeout or all events having been ACKed by outputs.
-		fb.sigWait.Add(withLog(wgEvents.Wait,
+	// Checks if on shutdown it should wait for all events to be published
+	waitPublished := fb.config.ShutdownTimeout > 0 || *once
+	if waitPublished {
+		// Wait for registrar to finish writing registry
+		waitEvents.Add(withLog(wgEvents.Wait,
 			"Continue shutdown: All enqueued events being published."))
-		fb.sigWait.Add(withLog(waitDuration(timeout),
-			"Continue shutdown: Time out waiting for events being published."))
+		// Wait for either timeout or all events having been ACKed by outputs.
+		if fb.config.ShutdownTimeout > 0 {
+			logp.Info("Shutdown output timer started. Waiting for max %v.", timeout)
+			waitEvents.Add(withLog(waitDuration(timeout),
+				"Continue shutdown: Time out waiting for events being published."))
+		}
```

Comment:
Add:
```
} else {
    waitEvents.AddChan(fb.done)
}
```

This will ensure CTRL-C will be handled immediately/properly if run-once without shutdown_timeout is configured. If `fb.done` has been closed already, `waitEvents.Wait` will return immediately. If `fb.done` is closed while waiting for `wgEvents.Wait`, shutdown will continue.